simulator/virtual_machine.go

```diff
@@ -1884,19 +1884,134 @@ func (vm *VirtualMachine) UpgradeVMTask(ctx *Context, req *types.UpgradeVM_Task)
 	body := &methods.UpgradeVM_TaskBody{}
 
 	task := CreateTask(vm, "upgradeVm", func(t *Task) (types.AnyType, types.BaseMethodFault) {
+
+		newInvalidStateFault := func(format string, args ...any) *types.InvalidState {
+			msg := fmt.Sprintf(format, args...)
+			return &types.InvalidState{
+				VimFault: types.VimFault{
+					MethodFault: types.MethodFault{
+						FaultCause: &types.LocalizedMethodFault{
+							Fault: &types.SystemErrorFault{
+								Reason: msg,
+							},
+							LocalizedMessage: msg,
+						},
+					},
+				},
+			}
+		}
+
+		// InvalidPowerState
+		//
+		// 1. Is VM's power state anything other than powered off?
+		if vm.Runtime.PowerState != types.VirtualMachinePowerStatePoweredOff {
+			return nil, &types.InvalidPowerStateFault{
+				ExistingState:  vm.Runtime.PowerState,
+				RequestedState: types.VirtualMachinePowerStatePoweredOff,
+			}
+		}
+
+		// InvalidState
+		//
+		// 1. Is host on which VM is scheduled in maintenance mode?
+		// 2. Is VM a template?
+		// 3. Is VM already the latest hardware version?
+		var (
+			ebRef                     *types.ManagedObjectReference
+			latestHardwareVersion     string
+			hostRef                   = vm.Runtime.Host
+			supportedHardwareVersions = map[string]struct{}{}
+		)
+		if hostRef != nil {
+			var hostInMaintenanceMode bool
+			ctx.WithLock(*hostRef, func() {
+				host := ctx.Map.Get(*hostRef).(*HostSystem)
+				hostInMaintenanceMode = host.Runtime.InMaintenanceMode
+				if host.Parent == nil {
+					panic("host.Parent is nil")
```

Comment:
Had you hit this case? Seems it'd be fine to just let the (unlikely?) panic happen in the switch below, rather than have an explicit host.Parent check here. I think Parent should only ever be nil for the root Folder.

Reply:
That was for testing. I can remove that.